### PR TITLE
feat(saml): Support IdP initiated SSO

### PR DESCRIPTION
**Pull request by evanpurkhiser**
In order to support IdP initiated SSO, the ACS endpoint must be aware of the organization the user is authenticating to. Becuase the generic /auth/sso endpoint expects the user to already have initiated the login flow through the org login page, this endpoint cannot work for a IdP initiated sign in.

This adds a SAML specific SSO endpoint that will proxy to either the SSO org login init action or the in-progress auth-sslo action.

### CODE REVIEW

File: src/sentry/auth/providers/saml2.py [hunk 1]
```diff
@@ -74,6 +74,36 @@ def dispatch(self, request, helper):
...
+            return sso_login.handle(request)
+
+        # IdP initiated authentication. Start from org login flow
+        from sentry.web.frontend.auth_organization_login import AuthOrganizationLoginView
+
+        # AuthOranizationLogin will init the login flow *only if* the ``init``
+        # parameter is set.
+        request.POST._mutable = True
+        request.POST['init'] = True
+        request.POST._mutable = False
```
Comment (EvanPurkhiser) [inline src/sentry/auth/providers/saml2.py]: I'm not super happy about this at all, so definitely open to ideas

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Changes require @getsentry/security sign-off</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

### Security concerns found

- src/sentry/auth/providers/saml2.py


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/auth/providers/saml2.py (hunk 1 above)
Comment (dcramer) [reply src/sentry/auth/providers/saml2.py]: 🤢

idk what else you'd do unless you wanted to subclass some base "AuthClass" that just dry'd up the code

Comment (EvanPurkhiser) [reply src/sentry/auth/providers/saml2.py]: There's not too much shared code between these views unfortunately, so it makes it difficult. I could break the `AuthOrganizationLoginView.handler` method into a separate method that takes an optional parameter to init the helper.

Comment (mattrobenolt) [reply src/sentry/auth/providers/saml2.py]: Can't you copy into a new object then mutate that object, then reassign back to request.POST?

fwiw, I don't particularly care. I'd just maybe add a comment explaining this at least so it's an explicit decision.

**Review by mattrobenolt — APPROVED** [review]
Seems ok. If you wanna do something different with request.POST, that's fine, but as is it's not terrible.

**Pull request merged**